internal/queue/pending_txs.go

```diff
@@ -24,30 +23,49 @@ func NewPendingTxs(logger *logger.SugarLogger) *PendingTxs {
 	}
 }
 
-func (p *PendingTxs) Add(txID string, promise *CompletionPromise) {
-	p.Lock()
-	defer p.Unlock()
+// Add returns true if the txId was already taken
+func (p *PendingTxs) Add(txID string, promise *CompletionPromise) bool {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	_, loaded := p.txs[txID]
+	if !loaded {
+		p.txs[txID] = promise
+	}
+	return loaded
+}
+
+func (p *PendingTxs) DeleteWithNoAction(txID string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	delete(p.txs, txID)
+}
 
-	p.txs[txID] = promise
+func (p *PendingTxs) loadAndDelete(txID string) (*CompletionPromise, bool) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	promise, loaded := p.txs[txID]
+	if loaded {
+		delete(p.txs, txID)
+	}
```

Comment:
Most of the time `loaded=true`. It is `false` only when a node is committing after a leadership change with pending TXs in the Raft queues which are not in the PendindTxs queue, as they were submitted in another node (which shutdown). This is extremely rare, so we might as well just delete without checking.